README.md
# Newsletters to Kindle
Uses IMAP to check a mailbox, converts HTML newsletter emails to PDFs and sends them to an Amazon Kindle email address

[![Latest Stable Version](https://poser.pugx.org/alexpcooper/newsletters-to-kindle/v/stable)](https://packagist.org/packages/alexpcooper/newsletters-to-kindle)
[![Total Downloads](https://poser.pugx.org/alexpcooper/newsletters-to-kindle/downloads)](https://packagist.org/packages/alexpcooper/newsletters-to-kindle)
[![License](https://poser.pugx.org/alexpcooper/newsletters-to-kindle/license)](https://packagist.org/packages/alexpcooper/newsletters-to-kindle)


### Process
1. Checks a folder in an email inbox, via IMAP
2. Collects the email and converts it into a PDF
3. Sends the PDF to Amazon, to convert into an ePub format and place in your Amazon Kindle library
4. Deletes the email


## Installation

Official installation method is via composer and its packagist package [alexpcooper/newsletters-to-kindle](https://packagist.org/packages/alexpcooper/newsletters-to-kindle).

```
$ composer require alexpcooper/newsletters-to-kindle
```


## Usage

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

$newsletter2kindle = new newsletters2kindle();
$newsletter2kindle->kindle_email = 'your-kindle-email@kindle.com';
$newsletter2kindle->imap_email   = 'your-email@address.com';
$newsletter2kindle->imap_user    = 'your-email@address.com';
$newsletter2kindle->imap_pass    = 'emailpassword';
$newsletter2kindle->imap_host    = 'imap.host.com';
$newsletter2kindle->imap_port    = 993;
$newsletter2kindle->imap_dir     = 'Inbox';

$newsletter2kindle->checkMail();

```

## Debug

As the process of sending an email to a 3rd party (Amazon) is "fire and forget", without any confirmation or failure, there are two optional parameters, in the event of needing to troubleshoot.

Add one or both of these prior to calling `checkMail()`.

```php
$newsletter2kindle->debug               = true; // false by default or when not specified
$newsletter2kindle->delete_mail_after   = false; // true by default or when not specified

```

### Parameters

| Parameter 	| Explanation 	|
|---	|---	|
| kindle_email 	| Used to send the PDF email to, so that it goes to your Amazon account 	|
| imap settings  	| Used to both check an email directory, such as your inbox, for an email to convert, as well as to send the PDF to Amazon as the FROM  address 	|
| imap_dir  	| This can be your inbox or a subfolder, eg. "Inbox", "[Gmail]/All Mail", "my-newsletters", etc.  	|
| debug 	| Prints out progress or any issues as it goes; for debugging only (defaults to false) 	|
| delete_mail_after 	| Keeps the mail in the mailbox after processing (defaults to true) - note that subsequent requests will pick up the same email again 	|


## Required Packages
With thanks to the following for these dependancies;
- https://github.com/mpdf/mpdf
- https://github.com/zbateson/mail-mime-parser
- https://github.com/swiftmailer/swiftmailer


## Troubleshooting
* Ensure that your "kindle_email" address is correct as it appears on your Amazon account for receiving documents (see [Edit Your Send to Kindle Email Address](https://www.amazon.co.uk/gp/help/customer/display.html?ref_=hp_left_v4_sib&nodeId=G7V489F2ZZU9JJGE))
* Ensure that your sending email address is permitted to send documents to your Amazon Kindle (see [Add an Email Address to Receive Documents in Your Kindle Library](https://www.amazon.co.uk/gp/help/customer/display.html?nodeId=GX9XLEVV8G4DB28H))


## Known Issues
- Due to the composition of emails, occasionally a block of style tags may be printed onto the end document. This is caused by external sources, such as includes / images, being pulled in after the email is parsed.
- At present the Kindle document appears in Amazon without an Author being populated, even though it's on the PDF. This seems to be caused on Amazon's side, when they convert the document.


CHANGELOG.md
# Newsletters to Kindle: ChangeLog


## Version 1.2.0
- Removed further script tags in the end document; added Title, Author and Creator to the PDF metadata
- Added further cleaning-up of script tags, removed emojis from subject / title

## Version 1.1.1
- Uncommented the deleting of the email
- Added known issues to ReadMe

## Version 1.1.0
- UTF encoding of PDF subject and author
- Creation of ReadMe

## v1.0.0 Pre-release


src/NewsletterToKindle.php
<?php 
    namespace AlexPCooper;

//    require('vendor/autoload.php');

    use ZBateson\MailMimeParser\MailMimeParser;
    use ZBateson\MailMimeParser\Header\HeaderConsts;
    use PHPMailer\PHPMailer\PHPMailer;
    use PHPMailer\PHPMailer\Exception;

    class NewsletterToKindle
    {
        protected $kindle_email;
        protected $imap_email;
        protected $imap_host;
        protected $imap_port;
        protected $imap_dir;
        protected $imap_user;
        protected $imap_pass;

        protected $debug;

        private $imap_conn;
        private $mail_message;
        private $pdf_document;
        private $mail_subject;

        function __construct($kindle_email, $imap_email, $imap_user, $imap_pass, $imap_host, $imap_port, $imap_dir, $debug = false) 
        {
            $this->kindle_email = $kindle_email;
            $this->imap_email   = $imap_email;
            $this->imap_user    = $imap_user;
            $this->imap_pass    = $imap_pass;
            $this->imap_host    = $imap_host;
            $this->imap_port    = $imap_port;
            $this->imap_dir     = $imap_dir;

            $this->debug        = $debug;

            $this->init();
        }

        public function init()
        {
            if ($this->checkMailbox())
            {
                $this->getMessage();
                $this->makePDF();
                $this->sendToKindle();

                imap_delete($this->imap_conn, 1);
            }

            if ($this->imap_conn)
            {
                imap_close($this->imap_conn); 
            }
        }

        private function checkMailbox()
        {
            $this->imap_conn = imap_open("{".trim($this->imap_host).":".trim($this->imap_port)."/imap/ssl}".$this->imap_dir, $this->imap_user, $this->imap_pass) or die('Failed to open connection: ' . imap_last_error());

            $msgs_number = 0;
            if ($this->imap_conn)
            {
                // get the total amount of messages
                $mbox_check = imap_check($this->imap_conn);
                $msgs_number = $mbox_check->Nmsgs;
            }

            return $msgs_number;
                
        }
        
        private function getMessage()
        {
            // get message body
            $imap_body = imap_fetchbody($this->imap_conn, 1, "");
    
            // clean up, as best as we can, the raw HTML / styling from the email body
            $imap_body = str_replace('<style type=3D"text/css">', '<style type="text/css">', $imap_body);
            $imap_body = strip_tags($imap_body, '<style><a><p>><br><br /><b><strong><u><i><em><img><h1><h2><h3><h4><h5><h6>');
            $imap_body = preg_replace('#<style type="text/css">.*?</style>#s', '', $imap_body);
    
            // create MailMimeParser
            $mail_parser = new MailMimeParser();
            $this->mail_message = $mail_parser->parse($imap_body, true);

            return true;
        }


        private function makePDF()
        {

            // use the email subject as the subject of the PDF
            $this->mail_subject = $this->mail_message->getHeaderValue(HeaderConsts::SUBJECT);
            if (strlen(trim($this->mail_subject)) == 0)
            {
                $this->mail_subject = 'Newsletter';
            }
            



            $this->pdf_document = new \Mpdf\Mpdf(['mode' => 'utf-8', 'format' => 'A4', 'default_font_size' => 25, 'orientation' => 'P', 'ignore_table_widths' => true, 'shrink_tables_to_fit' => false ]);
        
            // med-specific commands to ensure we get the best result in a PDF
            $this->pdf_document->shrink_tables_to_fit = 0;
            $this->pdf_document->SetDisplayMode('fullwidth');
            $this->pdf_document->allow_charset_conversion=true;
            $this->pdf_document->charset_in='UTF-8';

            $this->pdf_document->SetSubject($this->mail_subject); 

            $this->pdf_document->WriteHTML('<h1>'.$this->mail_subject.'</h1><h2>'.$this->mail_message->getHeader(HeaderConsts::FROM)->getPersonName().'</h2>');
            $this->pdf_document->AddPage();

            // chunk the email body because otherwise you can get issues with blank pages
            foreach (str_split($this->mail_message->getHtmlContent(), 10000) as $chunk)
            {
                $this->pdf_document->WriteHTML(mb_convert_encoding($chunk, "HTML-ENTITIES", "UTF-8"));
            }
    
            // $mpdf->Output(); die();

            return $this->pdf_document;
        }
        

        private function sendToKindle()
        {
            $mail = new PHPMailer(true);

            try 
            {
                // Mail Server settings
                $mail->Host       = $this->imap_host;
                $mail->SMTPAuth   = true;                                   //Enable SMTP authentication
                $mail->Username   = $this->imap_user;
                $mail->Password   = $this->imap_pass;
                $mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS;            //Enable implicit TLS encryption
                $mail->Port       = $this->imap_port;  
    
                //Recipients
                // $mail->setFrom($this->imap_email);
                $mail->addAddress($this->kindle_email);

                $mail->addAddress($this->imap_email);
    
                //Attachments
                $mail->AddStringAttachment( $this->pdf_document->Output('', 'S') , 'newsletter_'.str_replace(' ', '_', strtolower($this->mail_subject)).'.pdf', 'base64', 'application/pdf');
    
                //Content
                $mail->Subject = 'Convert'; // has to be this for ePub / Kindle format (or it provides it as PDF);
                $mail->Body    = $this->mail_subject;
                $mail->AltBody = $this->mail_subject;
    
                $mail->send();
    
                if ($this->debug)
                {
                    echo 'Message "'.$this->mail_subject.'" has been sent to '.$this->kindle_email;
                }
                return true;
            } 
            catch (\Exception $e) 
            {
                if ($this->debug)
                {
                    echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
                }
                return false;
            }
    

        }
    }

src/newsletters2kindle.php
<?php 
    namespace AlexPCooper;

//    require('vendor/autoload.php');

    use ZBateson\MailMimeParser\MailMimeParser;
    use ZBateson\MailMimeParser\Header\HeaderConsts;
    use PHPMailer\PHPMailer\PHPMailer;
    use PHPMailer\PHPMailer\Exception;

    class newsletters2kindle
    {
        public $kindle_email;
        public $imap_email;
        public $imap_host;
        public $imap_port;
        public $imap_dir;
        public $imap_user;
        public $imap_pass;

        public $debug;
        public $delete_mail_after;

        private $imap_conn;
        private $mail_message;
        private $pdf_document;
        private $mail_subject;
        private $mail_from;

        function __construct() 
        {
            $this->debug                = false;
            $this->delete_mail_after    = true;
        }

        public function checkMail()
        {
            if ($this->checkMailbox())
            {
                $this->getMessage();
                $this->makePDF();
                $this->sendToKindle();

                if ($this->delete_mail_after)
                {
                    imap_delete($this->imap_conn, 1);
                }
            }
            elseif ($this->debug)
            {
                echo 'No newsletters / emails found';
            }

            if ($this->imap_conn)
            {
                imap_close($this->imap_conn); 
            }
        }

        private function checkMailbox()
        {
            $this->imap_conn = imap_open("{".trim($this->imap_host).":".trim($this->imap_port)."/imap/ssl}".$this->imap_dir, $this->imap_user, $this->imap_pass) or die('Failed to open connection: ' . imap_last_error());

            $msgs_number = 0;
            if ($this->imap_conn)
            {
                // get the total amount of messages
                $mbox_check = imap_check($this->imap_conn);
                $msgs_number = $mbox_check->Nmsgs;
            }

            return $msgs_number;
                
        }
        
        private function getMessage()
        {
            // get message body
            $imap_body = imap_fetchbody($this->imap_conn, 1, "");
    
            // clean up, as best as we can, the raw HTML / styling from the email body
            $imap_body = str_replace('type=3D"text"', 'type="text', $imap_body);
            $imap_body = str_replace('<style type=3D"text/css">', '<style type="text/css">', $imap_body);
            $imap_body = str_replace('<style amf:inline=3D"amf:inline" type=3D"text/css">', '<style type="text/css">', $imap_body);
            $imap_body = str_replace('<style amf:inline="amf:inline" type="text/css">', '<style type="text/css">', $imap_body);
            $imap_body = strip_tags($imap_body, '<style><a><p>><br><br /><b><strong><u><i><em><img><h1><h2><h3><h4><h5><h6>');
            $imap_body = preg_replace('#<style type="text/css">.*?</style>#s', '', $imap_body);
    
            // create MailMimeParser
            $mail_parser = new MailMimeParser();
            $this->mail_message = $mail_parser->parse($imap_body, true);

            // var_dump($imap_body); die();
            // var_dump($this->mail_message->getHtmlContent()); die();

            return true;
        }


        private function makePDF()
        {

            // use the email subject as the subject of the PDF
            $this->mail_subject = trim(utf8_encode(imap_utf8($this->replace_4byte($this->mail_message->getHeaderValue(HeaderConsts::SUBJECT)))));
            if (strlen(trim($this->mail_subject)) == 0)
            {
                $this->mail_subject = 'Newsletter';
            }

            $this->mail_from = utf8_encode(imap_utf8($this->mail_message->getHeader(HeaderConsts::FROM)->getPersonName()));
            if (strlen(trim($this->mail_from)) == 0)
            {
                $this->mail_from = imap_utf8($this->mail_message->getHeader(HeaderConsts::FROM));
            }
            $this->mail_from = trim(str_replace('From: ', '', $this->mail_from));



            $this->pdf_document = new \Mpdf\Mpdf(['mode' => 'utf-8', 'format' => 'A4', 'default_font_size' => 25, 'orientation' => 'P', 'ignore_table_widths' => true, 'shrink_tables_to_fit' => false ]);
        
            // med-specific commands to ensure we get the best result in a PDF
            $this->pdf_document->shrink_tables_to_fit = 0;
            $this->pdf_document->SetDisplayMode('fullwidth');
            $this->pdf_document->allow_charset_conversion=true;
            $this->pdf_document->charset_in='UTF-8';

            $this->pdf_document->SetSubject($this->mail_subject); 
            $this->pdf_document->SetTitle($this->mail_subject);
            $this->pdf_document->SetAuthor($this->mail_from);
            $this->pdf_document->SetCreator('Newsletters To Kindle');

            $this->pdf_document->WriteHTML('<h1>'.$this->mail_subject.'</h1><h2>'.utf8_encode($this->mail_from).'</h2>');
            $this->pdf_document->AddPage();

            // chunk the email body because otherwise you can get issues with blank pages
            foreach (str_split($this->mail_message->getHtmlContent(), 10000) as $chunk)
            {
                $this->pdf_document->WriteHTML(mb_convert_encoding($chunk, "HTML-ENTITIES", "UTF-8"));
            }
    
            // $this->pdf_document->Output(); die();

            return $this->pdf_document;
        }
        

        private function sendToKindle()
        {
            $mail = new PHPMailer(true);

            try 
            {
                // Mail Server settings
                $mail->Host       = $this->imap_host;
                $mail->SMTPAuth   = true;                                   //Enable SMTP authentication
                $mail->Username   = $this->imap_user;
                $mail->Password   = $this->imap_pass;
                $mail->SMTPSecure = PHPMailer::ENCRYPTION_SMTPS;            //Enable implicit TLS encryption
                $mail->Port       = $this->imap_port;  
    
                //Recipients
                $mail->setFrom($this->imap_email);
                $mail->addAddress($this->kindle_email);

                if ($this->debug)
                {
                    $mail->addAddress($this->imap_email);
                }

                //Attachments
                $mail->AddStringAttachment( $this->pdf_document->Output('', 'S') , utf8_encode($this->mail_subject).'.pdf', 'base64', 'application/pdf');
    
                //Content
                $mail->Subject = 'Convert'; // has to be this for ePub / Kindle format (or it provides it as PDF);
                $mail->Body    = $this->mail_subject;
                $mail->AltBody = $this->mail_subject;
    
                $mail->send();
    
                if ($this->debug)
                {
                    echo 'Message "'.$this->mail_subject.'" has been sent to '.$this->kindle_email;
                }
                return true;
            } 
            catch (Exception $e) 
            {
                if ($this->debug)
                {
                    echo "Message could not be sent. Mailer Error: {$mail->ErrorInfo}";
                }
                return false;
            }
    

        }


        // to remove 4byte characters like emojis etc..
        // https://stackoverflow.com/questions/12807176/php-writing-a-simple-removeemoji-function
        private function replace_4byte($string) 
        {
            return preg_replace('%(?:
                  \xF0[\x90-\xBF][\x80-\xBF]{2}      # planes 1-3
                | [\xF1-\xF3][\x80-\xBF]{3}          # planes 4-15
                | \xF4[\x80-\x8F][\x80-\xBF]{2}      # plane 16
            )%xs', '', $string);    
        }
    }